app/src/main/kotlin/compiler/analysis/InitializationVerifier.kt

```diff
@@ -0,0 +1,200 @@
+package compiler.analysis
+
+import compiler.ast.AstNode
+import compiler.ast.Expression
+import compiler.ast.Expression.BinaryOperation.Kind.AND
+import compiler.ast.Expression.BinaryOperation.Kind.OR
+import compiler.ast.Function
+import compiler.ast.NamedNode
+import compiler.ast.Program
+import compiler.ast.Program.Global
+import compiler.ast.Statement
+import compiler.ast.Variable
+import compiler.diagnostics.Diagnostic.ResolutionDiagnostic.VariableInitializationError.ReferenceToUninitializedVariable
+import compiler.diagnostics.Diagnostics
+import compiler.utils.Ref
+import compiler.utils.mutableRefSetOf
+
+object InitializationVerifier {
+    data class VerificationState(
+        val initializedVariables: MutableSet<Ref<NamedNode>> = mutableRefSetOf(),
+        val traversedFunctions: MutableSet<Ref<Function>> = mutableRefSetOf()
+    ) {
+        fun copy(): VerificationState = VerificationState(initializedVariables.toMutableSet(), traversedFunctions.toMutableSet())
+        fun add(other: VerificationState) {
+            initializedVariables.addAll(other.initializedVariables)
+            traversedFunctions.addAll(other.traversedFunctions)
+        }
+        fun restore(state: VerificationState) {
+            initializedVariables.clear()
+            initializedVariables.addAll(state.initializedVariables)
+            traversedFunctions.clear()
+            traversedFunctions.addAll(state.traversedFunctions)
+        }
+        infix fun intersect(other: VerificationState) = VerificationState(
+            (initializedVariables intersect other.initializedVariables).toMutableSet(),
+            (traversedFunctions intersect other.traversedFunctions).toMutableSet()
+        )
+    }
+
+    fun verifyAccessedVariablesAreInitialized(
+        ast: Program,
+        nameResolution: Map<Ref<AstNode>, Ref<NamedNode>>,
+        defaultParameterMapping: Map<Ref<Function.Parameter>, Variable>,
+        diagnostics: Diagnostics
+    ) {
+        fun verifyInitialization(
+            node: AstNode,
+            state: VerificationState
+        ): VerificationState? {
+            // Consider a function f, which performs g(). g might contain conditional returns.
```

Comment:
Great catch with these conditional returns. I think we should also treat `yield` as a possible conditional return, since `foreach` loop can be exited using `break` statement.